### PR TITLE
Pickle instead of scatter.

### LINKED ISSUE #302: Explicitly write to disk instead of `client.scatter`
**Bug report**

Troy has been running into issues with slow worker initialization causing problems with the `client.scatter` call. Because the workers are not yet visible to the main runner, the `client.scatter` fails. This is a known issue with dask distributed (https://github.com/dask/distributed/issues/2941). We can explicitly write out the large side inputs to disk (or pickle them), and send along file references that can be decoded on the workers.

This is more of a problem with SLURM distributed clusters than typical local cluster configurations.

**Before submitting**
Please check the following:

- [x] I have described the situation in which the bug arose, including what code was executed, information about my environment, and any applicable data others will need to reproduce the problem.
- [ ] I have included available evidence of the unexpected behavior (including error messages, screenshots, and/or plots) as well as a descriprion of what I expected instead.
- [x] If I have a solution in mind, I have provided an explanation and/or pseudocode and/or task list.

### PR DESCRIPTION
**Pull request by delucchi-cmu**
## Change Description

Closes #302.

Writes the file reader to a pickled file, and de-pickles in the mapping and splitting workers.

Writes the alignment to a pickled file, and de-pickles in the splitting workers. destination map is created on-demand, and is owned by the resume plan, as this is used to determine remaining reduce tasks.

## Code Quality
- [x] I have read the [Contribution Guide](https://hipscat-import.readthedocs.io/en/stable/guide/contributing.html) and [LINCC Frameworks Code of Conduct](https://lsstdiscoveryalliance.org/programs/lincc-frameworks/code-conduct/)
- [x] My code follows the code style of this project
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

### CODE REVIEW

File: src/hipscat_import/catalog/resume_plan.py [hunk 1]
```diff
@@ -196,6 +200,61 @@ def is_mapping_done(self) -> bool:
...
+            with open(file_name, "ab") as pickle_file:
+                pickle.dump(alignment, pickle_file)
+
+        if self.destination_pixel_map is None:
+            with open(file_name, "rb") as pickle_file:
+                alignment = pickle.load(pickle_file)
+            non_none_elements = alignment[alignment != np.array(None)]
+            self.destination_pixel_map = np.unique(non_none_elements)
+            self.destination_pixel_map = [
+                (pix, order, count) for (pix, order, count) in self.destination_pixel_map if int(count) > 0
```
Comment (camposandro) [inline src/hipscat_import/catalog/resume_plan.py:253]: The alignment seems to return the order, then pixel number:
```suggestion
                (order, pix, count) for (order, pix, count) in self.destination_pixel_map if int(count) > 0
```

File: src/hipscat_import/catalog/resume_plan.py [hunk 2]
```diff
@@ -208,30 +267,31 @@ def is_splitting_done(self) -> bool:
...
-            self.destination_pixel_map = destination_pixel_map
         if self.destination_pixel_map is None:
             raise RuntimeError("destination pixel map not provided for progress tracking.")
         reduce_items = [
-            (hp_pixel, source_pixels, f"{hp_pixel.order}_{hp_pixel.pixel}")
-            for hp_pixel, source_pixels in destination_pixel_map.items()
-            if f"{hp_pixel.order}_{hp_pixel.pixel}" not in reduced_keys
+            (HealpixPixel(hp_pixel, hp_order), row_count, f"{hp_pixel}_{hp_order}")
+            for hp_pixel, hp_order, row_count in self.destination_pixel_map
+            if f"{hp_pixel}_{hp_order}" not in reduced_keys
```
Comment (camposandro) [inline src/hipscat_import/catalog/resume_plan.py:285]: I think the pixel initialization takes in the order first: 
```suggestion
            (HealpixPixel(hp_order, hp_pixel), row_count, f"{hp_order}_{hp_pixel}")
```

```suggestion
            (HealpixPixel(hp_order, hp_pixel), row_count, f"{hp_order}_{hp_pixel}")
            for hp_order, hp_pixel, row_count in self.destination_pixel_map
            if f"{hp_order}_{hp_pixel}" not in reduced_keys
```

File: src/hipscat_import/catalog/resume_plan.py [hunk 3]
```diff
@@ -208,30 +267,31 @@ def is_splitting_done(self) -> bool:
...
         ]
         return reduce_items
 
+    def get_destination_pixels(self):
+        """Create HealpixPixel list of all destination pixels."""
+        if self.destination_pixel_map is None:
+            raise RuntimeError("destination pixel map not known.")
+        return [HealpixPixel(hp_pixel, hp_order) for hp_pixel, hp_order, _ in self.destination_pixel_map]
```
Comment (camposandro) [inline src/hipscat_import/catalog/resume_plan.py:293]:
```suggestion
        return [HealpixPixel(hp_order, hp_pixel) for hp_order, hp_pixel, _ in self.destination_pixel_map]
```

File: src/hipscat_import/catalog/resume_plan.py [hunk 4]
```diff
@@ -196,6 +200,61 @@ def is_mapping_done(self) -> bool:
...
+                )
+            with open(file_name, "ab") as pickle_file:
+                pickle.dump(alignment, pickle_file)
+
+        if self.destination_pixel_map is None:
+            with open(file_name, "rb") as pickle_file:
+                alignment = pickle.load(pickle_file)
+            non_none_elements = alignment[alignment != np.array(None)]
+            self.destination_pixel_map = np.unique(non_none_elements)
+            self.destination_pixel_map = [
```
Comment (camposandro) [inline src/hipscat_import/catalog/resume_plan.py:252]: Does this `destination_pixel_map` comply with the specified type?
```
destination_pixel_map: Optional[List[Tuple[HealpixPixel, List[HealpixPixel], str]]]
```

Comment (delucchi-cmu) [reply src/hipscat_import/catalog/resume_plan.py:252]: Nope! Fixed.

File: src/hipscat_import/catalog/resume_plan.py (hunk 1 above)
Comment (delucchi-cmu) [reply src/hipscat_import/catalog/resume_plan.py]: Luckily, this was just an issue of naming the variables incorrectly, but I was doing in consistently, so it still had the right behavior!!

File: src/hipscat_import/catalog/resume_plan.py (hunk 2 above)
Comment (delucchi-cmu) [reply src/hipscat_import/catalog/resume_plan.py]: Same as above, with bad (but consistent) naming.

**Review by camposandro — APPROVED** [review]

**Pull request merged**